extensions/arc/deployment/src/main/java/io/quarkus/arc/deployment/ContextRegistrarBuildItem.java

```diff
@@ -1,20 +1,50 @@
 package io.quarkus.arc.deployment;
 
+import java.lang.annotation.Annotation;
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Collection;
+
+import org.jboss.jandex.DotName;
+
 import io.quarkus.arc.processor.ContextRegistrar;
 import io.quarkus.builder.item.MultiBuildItem;
 
 /**
- * Make it possible to register a custom CDI context.
+ * Make it possible to register one or more custom CDI contexts.
+ * If you are registering a new context, you should also pass in the annotation value into the constructor either in
+ * {@link DotName} form, or as {@code Class<? extends Annotation>}.
+ *
+ * This information is then leveraged in {@link CustomScopeAnnotationsBuildItem} which allows consumers to browse
+ * all known custom scoped within deployment even early in the build process.
  */
 public final class ContextRegistrarBuildItem extends MultiBuildItem {
 
     private final ContextRegistrar contextRegistrar;
+    private final Collection<DotName> annotationNames;
 
-    public ContextRegistrarBuildItem(ContextRegistrar contextRegistrar) {
+    public ContextRegistrarBuildItem(ContextRegistrar contextRegistrar, DotName... annotationsNames) {
         this.contextRegistrar = contextRegistrar;
+        if (annotationsNames == null) {
+            throw new IllegalArgumentException("Parameter annotationNames cannot be null!");
```

Comment:
Hm, so if I pass an empty array it's OK? Also if you skip the `annotationsNames` completely on the call site the method receives an empty array. It could be compiler-specific though. This is what I get for oracle JDK 8.

Anyway, I'm not quite sure whether throwing an IAE is the best approach. I'd rather just log a warning if a `ContextRegistrarBuildItem` has no annotations. In theory, the `ContextRegistrar` can register a context conditionally... I know that it's ugly :-(.

Reply:
https://stackoverflow.com/questions/10905942/which-part-of-the-java-language-specification-describes-the-behaviour-of-omitted